### PR TITLE
Premium Licence Implementation

### PR DESCRIPTION
**Pull request by EmranMR**
Hi @mikebronner 👋

Hope you are well! So the Intelephense **PREMIUM** features are in fact available in Nova (not all! but most!) so I tried to implement it however I think I might need a bit of your help to make it bug free please see below:

**Intelephense Premium Activation:**
---
- So based on the [intelephense premium installation guide](https://intelephense.com/faq.html)
- User needs to have the licence code saved in a file at ~/intelephense/licence.txt 
- This is only for third party IDEs, I am not sure how the VSCode works because when we put the licence in, it does not write a file in that directory

**Premium Features in Nova**
---
- I personally had this file in the directory manually and worked okay ish with both intelephense extensions including Kristófer's
- As I mentioned above not all features are available, I tried to do an audit to see what's available and what is not but as far as I know the issue was Nova's lack of full support for LSP (I am not an expert in LSP by any means, maybe you could also add a feature request +1 as I think I emailed them myself as well) _renaming symbols_ is one such examples not working
- None the less some of the other very useful features work
- Some of the main features are  accessed with "code action" `command + shift + A` for example to add PHPDoc etc 
Screenshot 2022-07-13 at 11 31 55

- I was supposed to do a pull request to Kristófer extension, but decided to start with this one as its hosted on GitHub and easier for me at the moment. 
- if done manually this is acknowledged in the extension console 

Screenshot 2022-07-13 at 10 48 54

**The Pull Request**
---
- Sorry in advance I had prettier on and the the code style change messed up with the diffs so I will explain below
- in `export.activation` it is checked if licence.txt already exists, if so it disables the command by writing false in the config and vice versa
- I added a new command under yours for _restart intelephense_, called _enter intelephense licence_
- Everything is done via imported `Licence` class
- it asks for users number, takes that and tries to write in to the licence.txt

**Problems**
---
- I realised if I use `restart()` it does not change the confitg file values set by `config.set` in `export.activation` 
- my understating is that restart does deactivate and activate again? might be just API limitation so I am not sure about a workaround to fully restart the extension once the licence is written in. Otherwise the user has to close and open nova for the premium features to take effect.
- Even manual restart doesn't work by using the command you have created from command palette
- I had to use `try/catch` in the `registerLicence` method because once the file is created and the user tries to restart, the extension console would give an error, this way at least it is ignored. Again simply because the config file is not changed for some reason during activate/deactivate..

**☠️ Death Bug ☠️**
---
- There is a bug that makes Bricks Nova 🧱  ☠️⚰️
- This has nothing to do with this pull request
- It is more likely to be Nova app problem so maybe we should file this bug with them?
- basically if the user has mistype or write the **wrong** license code in the licence.txt this would gives a notification in Nova that there is such problem as shown below. Soon after Nova dies 🛑
- The only way for user to fix would be to go and change the file or delete it manually otherwise Nova app would be fully bricked
Screenshot 2022-07-13 at 09 52 45

Sorry for the long message and if you require further clarification about the codes do get in touch I hope you find these helpful!
Thanks
Emran

### CODE REVIEW

**Comment by mikebronner** [comment]
Love it, thanks for all your hard work! What if we cleared the incorrect license key if activation fails, would that prevent the bricking? Leave a comment when you are done working on this and I will take a look and make any necessary adjustments.

**Comment by EmranMR** [comment]
> Love it, thanks for all your hard work! What if we cleared the incorrect license key if activation fails, would that prevent the bricking? Leave a comment when you are done working on this and I will take a look and make any necessary adjustments.

That is actually a very good workaround, However I am totally stuck on how to intercept the error. Firstly I don’t know how to access the intelephense JSON response that gets logged in to the console by the LSP.  One other major problem I have is, I don’t even get a chance to read anything in the console. The app crashes very rapidly as soon as the invalid licence notification pops up haha! 😬

I have however logged this bug with Nova (not on the forum) and hopefully someone will get back to me. 🤞
